FeatureEngine.py

```diff
@@ -42,37 +44,243 @@ def map_labels_to_invoice_OCR(self, data_path: str, verbose: bool = False):
+        def is_vert_aligned(t1, t2, moe):
+            if abs(t1.coordinates["x"] - t2.coordinates["x"]) < moe:
+                return True
+            if (
+                abs(
+                    (t1.coordinates["x"] + t1.coordinates["width"])
+                    - (t2.coordinates["x"] + t2.coordinates["width"])
+                )
+                < moe
+            ):
+                return True
+            t1_midpt_x = t1.coordinates["x"] + (t1.coordinates["width"] / 2)
+            t2_midpt_x = t2.coordinates["x"] + (t2.coordinates["width"] / 2)
+            if abs(t1_midpt_x - t2_midpt_x) < moe:
+                return True
+            return False
+
+        # checks if two tokens are aligned vertically within a margin of error (checks midpoint, top boundary, bottom boundary)
+        def is_hori_aligned(t1, t2, moe):
+            if abs(t1.coordinates["y"] - t2.coordinates["y"]) < moe:
+                return True
+            if (
+                abs(
+                    (t1.coordinates["y"] + t1.coordinates["height"])
+                    - (t2.coordinates["y"] + t2.coordinates["height"])
+                )
+                < moe
+            ):
+                return True
+            t1_midpt_y = t1.coordinates["y"] + (t1.coordinates["height"] / 2)
+            t2_midpt_y = t2.coordinates["y"] + (t2.coordinates["height"] / 2)
+            if abs(t1_midpt_y - t2_midpt_y) < moe:
+                return True
+            return False
+
+        #############################################################################################################################
+
         features = {}
 
-        ### Datatype-related features ###
-        # TODO: Implement regexes to determine datatype of this token - DATE, MONEY, NUMBER
+        # number of characters and words in token text string
+        features["char_count"] = len(token.text)
+        features["word_count"] = len(token.text.split(" "))
 
-        ### Vertical location feature ###
-        # TODO: Calculate distance from the top of the image
+        # height and width of token
+        features["height"] = token.coordinates["height"]
+        features["width"] = token.coordinates["width"]
 
-        ### Date-alignment related features ###
-        date_related_tokens = invoicePage.search_tokens("date")
+        # distance to edges of page (to nearest point on the box)
+        features["dist_top"] = token.coordinates["y"]
+        features["dist_left"] = token.coordinates["x"]
+        features["dist_bottom"] = invoicePage.size["y"] - (
+            token.coordinates["y"] + token.coordinates["height"]
+        )
+        features["dist_right"] = invoicePage.size["x"] - (
+            token.coordinates["x"] + token.coordinates["width"]
+        )
 
-        # Feature 1: Alignment with "date"
-        features["aligned_with_date"] = self.create_alignment_feature(
-            token, date_related_tokens
+        # relative size of token box compared to page
```

Comment:
Again, i'm not sure if you want to consider looking at the relative size of token box compared to the area of the page with text or pics etc